File: src/core/age/keys.ts
/**
 * age 私钥的生成与落盘（docs/m3-plan.md §2.2 / §1-D2，P1-W1 落地）。
 *
 * 布局（§1-D2，冻结）：`<keysDir>/age.txt`，即 `<home>/keys/age.txt`，
 * 权限 **0600**，幂等由 `ensureGitRepo` 的 `.gitignore` 的 `keys/` 行保证**不入库**。
 *
 * ## 安全不变量（本文件的全部设计围绕它）
 *
 * 1. **私钥永不进错误消息 / stdout / 日志**：本文件所有 `CliError` 的文案都是**固定文本**，
 *    绝不拼接文件内容或私钥。唯一的例外通道是把 `identity` 原样返回给调用方（内存内使用）。
 * 2. **绝不覆盖已有私钥**：`writeIdentityFile` 见到已存在文件一律 `CliError`——覆盖等于
 *    永久销毁旧机可解的历史密文（`secrets/` 里已推送的 `.age` 文件）。
 * 3. **原子写**：`<file>.tmp-<pid>` + `rename`（同目录 rename 在同一文件系统上原子），
 *    避免写到一半崩溃留下被 `loadIdentity` 读成「损坏 → undefined」的半个私钥。
 * 4. **读路径不 throw**：`loadIdentity` 是 keygen/push/pull/doctor/home 的共用读口，
 *    缺失或损坏都降级为 `undefined`，由调用方决定是「提示 keygen」还是「警告」——
 *    与 `src/core/state.ts` 的 `loadState` 同款容错策略。
 */

import fs from 'node:fs';
import path from 'node:path';

import { CliError } from '../errors.js';
import type { HomerPaths } from '../paths.js';
import { identityFromSecretKey, newAgeIdentity } from './cipher.js';
import type { AgeIdentity } from './types.js';

/**
 * §2.2 把 `recipientIsValid` 列在 keys.ts 名下，但它的**权威实现**在 `types.ts`
 * （P0 落地：`src/core/config.ts` 的配置校验需要它，若定义在此会把 config 拖进
 * 密码学依赖）。此处**原样 re-export**，使「从 keys.js 取该校验函数」的调用方
 * （§2.2 字面读法）也能工作，同时保持实现唯一一份、零重复。
 * `index.ts` 对同名导出只保留一条路径，避免 `export *` 歧义。
 */
export { recipientIsValid } from './types.js';

/** identity 文件名（§1-D2 冻结：`<home>/keys/age.txt`）。 */
const IDENTITY_FILE_NAME = 'age.txt';

/**
 * 生成一对新 age X25519 identity（**纯内存，同步，不落盘**）。
 *
 * §2.2 冻结签名为同步；实现见 `cipher.ts`（`age-encryption` 的 `generateIdentity()`
 * 是 async，但其内部逻辑同步，故这里用 `node:crypto` 直接做，见该文件头注释）。
 * 生成的 `secretKey` 只回给调用方，永不进错误消息 / 日志。
 */
export function generateIdentity(): AgeIdentity {
  return newAgeIdentity();
}

/** 本机 identity 文件路径（`<keysDir>/age.txt`）。不做存在性检查、不创建目录。 */
export function identityFilePath(paths: HomerPaths): string {
  return path.join(paths.keysDir, IDENTITY_FILE_NAME);
}

/**
 * 解析 identity 文件内容 → `AgeIdentity`。
 *
 * 接受 age 官方 key 文件的最小形态：**首个非空、非 `#` 注释行**即私钥（与 `age` CLI 的
 * `-i` 文件格式兼容，便于将来手工从 age CLI 迁移）。其余行只允许空行 / 注释；出现第二个
 * 私钥行 → `CliError`（`AgeIdentity` 是单密钥模型，静默丢弃一个私钥会让用户以为生效了）。
 *
 * 非法输入（空内容 / 前缀不符 / 长度不对 / checksum 错 / 混用大小写 / 多密钥）一律
 * `CliError`，且**消息绝不含内容本体**（内容通常就是私钥）。
 */
export function parseIdentityFile(content: string): AgeIdentity {
  const lines = content.split(/\r?\n/);
  const keyLines: string[] = [];
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed === '' || trimmed.startsWith('#')) continue;
    keyLines.push(trimmed);
  }

  if (keyLines.length === 0) {
    throw new CliError('age identity 文件为空：期望一行 `AGE-SECRET-KEY-1...`');
  }
  if (keyLines.length > 1) {
    // 只报行数，不回显任何一行内容。
    throw new CliError(
      `age identity 文件含 ${keyLines.length} 行私钥：homer 只支持单个 identity（请只保留一个）`,
    );
  }

  // identityFromSecretKey 内部做前缀 / 长度 / checksum 校验，且错误消息不含私钥本体。
  return identityFromSecretKey(keyLines[0]!);
}

/** identity 文件的文本形态：私钥一行 + 末尾换行（与 age CLI 的 key 文件一致）。 */
function serializeIdentity(identity: AgeIdentity): string {
  return `${identity.secretKey}\n`;
}

/**
 * 写入 `<keysDir>/age.txt`（**0600**，原子写，**拒绝覆盖**）。
 *
 * 顺序刻意如此：先检查存在性（并发窗口靠 `rename` 前的存在性判断 + 写前再查一次收窄，
 * 但本命令是单进程 CLI，不做额外锁——见 §0 Non-goals 的 `homer unlock`/并发锁 v2）。
 * `keysDir` 以 0700 创建（目录权限收紧，防同机其他用户列举）。
 */
export function writeIdentityFile(paths: HomerPaths, identity: AgeIdentity): void {
  const file = identityFilePath(paths);

  if (fs.existsSync(file)) {
    throw new CliError(
      `age identity 已存在，拒绝覆盖: ${file}`,
      '若要轮换密钥：先备份该文件并确认 secrets/ 里的密文已可被新 recipient 解开；homer 不提供覆盖写入',
    );
  }

  // 写入前再确认一次私钥合法（防调用方传入手搓对象把垃圾写进 keys/age.txt）。
  const canonical = identityFromSecretKey(identity.secretKey);

  fs.mkdirSync(paths.keysDir, { recursive: true, mode: 0o700 });

  const tmp = `${file}.tmp-${process.pid}`;
  try {
    // mode 传给 open(2)：受 umask 影响只会更严（0600 & ~umask ⊆ 0600），不会更宽。
    fs.writeFileSync(tmp, serializeIdentity(canonical), { encoding: 'utf8', mode: 0o600 });
    fs.chmodSync(tmp, 0o600); // 显式再设一次：对抗"用户 umask 异常"或既有 tmp 文件
    fs.renameSync(tmp, file);
  } catch (err) {
    fs.rmSync(tmp, { force: true }); // 失败不留下半个私钥
    throw err instanceof CliError
      ? err
      : new CliError(`写入 age identity 失败: ${file}（${err instanceof Error ? err.message : '未知错误'}）`);
  }
}

/**
 * 读取本机 identity：缺失 / 不可读 / 损坏 → `undefined`（**不 throw**）。
 *
 * 「损坏」包括：内容非法、多密钥、权限/IO 错误——读路径的职责是让调用方拿到一个
 * 可判定的「有没有」，而不是让整个 CLI 崩掉（对应 §2.6 push/pull 的 `no-identity` 状态、
 * doctor 的 age 检查、home 的 `secrets.skipped`）。
 */
export function loadIdentity(paths: HomerPaths): AgeIdentity | undefined {
  let content: string;
  try {
    content = fs.readFileSync(identityFilePath(paths), 'utf8');
  } catch {
    return undefined;
  }
  try {
    return parseIdentityFile(content);
  } catch {
    return undefined;
  }
}

File: internal/agecrypto/keys.go
package agecrypto

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"

	"filippo.io/age"

	"github.com/zzjcool/homer-cli/internal/core"
)

const identityFileName = "age.txt"

// AgeIdentity is the local age X25519 identity and its corresponding public
// recipient. SecretKey is only kept in memory and in the identity file under
// the user's Homer home.
type AgeIdentity struct {
	SecretKey string
	Recipient string
}

// keysDirectory returns the configured keys directory. HomerPaths produced by
// core.GetHomerPaths always have KeysDir populated, while accepting a partially
// populated path value keeps this package convenient for focused tests and
// callers constructing paths manually.
func keysDirectory(p core.HomerPaths) string {
	if p.KeysDir != "" {
		return p.KeysDir
	}
	if p.Home != "" {
		return filepath.Join(p.Home, "keys")
	}
	return "keys"
}

// IdentityFilePath returns the path of the local age identity file. It does
// not create the directory or inspect the filesystem.
func IdentityFilePath(p core.HomerPaths) string {
	return filepath.Join(keysDirectory(p), identityFileName)
}

// GenerateIdentity makes a new X25519 identity in memory. The age library
// only fails here when the system random source fails; keep the frozen
// synchronous API and return the zero value in that exceptionally rare case.
func GenerateIdentity() AgeIdentity {
	identity, err := age.GenerateX25519Identity()
	if err != nil {
		return AgeIdentity{}
	}
	return AgeIdentity{
		SecretKey: identity.String(),
		Recipient: identity.Recipient().String(),
	}
}

// KeyGen generates and persists a new identity. Existing identity files are
// never replaced because doing so would strand already-pushed vault files.
func KeyGen(p core.HomerPaths) (AgeIdentity, error) {
	identity := GenerateIdentity()
	if identity.SecretKey == "" || identity.Recipient == "" {
		return AgeIdentity{}, invalidIdentityError()
	}
	if err := WriteIdentityFile(p, identity); err != nil {
		return AgeIdentity{}, err
	}
	return identity, nil
}

// normalizeSecretKey accepts the all-lowercase spelling permitted by bech32
// and normalizes it to age's official all-uppercase identity spelling. Mixed
// case is deliberately rejected by the age format.
func normalizeSecretKey(secretKey string) (string, error) {
	if secretKey == "" || strings.TrimSpace(secretKey) != secretKey {
		return "", invalidIdentityError()
	}
	if secretKey != strings.ToLower(secretKey) && secretKey != strings.ToUpper(secretKey) {
		return "", invalidIdentityError()
	}
	candidate := secretKey
	if secretKey == strings.ToLower(secretKey) {
		candidate = strings.ToUpper(secretKey)
	}
	identity, err := age.ParseX25519Identity(candidate)
	if err != nil {
		return "", invalidIdentityError()
	}
	return identity.String(), nil
}

// parseSecretKey parses and derives an identity while ensuring all failures
// use a fixed error that cannot disclose the supplied private key.
func parseSecretKey(secretKey string) (*age.X25519Identity, error) {
	canonical, err := normalizeSecretKey(secretKey)
	if err != nil {
		return nil, err
	}
	identity, err := age.ParseX25519Identity(canonical)
	if err != nil {
		return nil, invalidIdentityError()
	}
	return identity, nil
}

func invalidIdentityError() error {
	return core.NewCliError("invalid age identity")
}

func duplicateIdentityError() error {
	return core.NewCliError("age identity already exists; refusing to overwrite")
}

// ParseIdentityFile parses the age CLI-compatible identity file format. It
// accepts one non-empty, non-comment line and rejects additional key lines.
// The returned identity is canonicalized to age's official spelling.
func ParseIdentityFile(content string) (AgeIdentity, error) {
	lines := strings.Split(strings.ReplaceAll(content, "\r\n", "\n"), "\n")
	var key string
	keyLines := 0
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		keyLines++
		if keyLines == 1 {
			key = line
		}
	}
	if keyLines == 0 || keyLines > 1 {
		return AgeIdentity{}, core.NewCliError(fmt.Sprintf("invalid age identity file: expected one key, found %d", keyLines))
	}

	identity, err := parseSecretKey(key)
	if err != nil {
		return AgeIdentity{}, err
	}
	return AgeIdentity{
		SecretKey: identity.String(),
		Recipient: identity.Recipient().String(),
	}, nil
}

// WriteIdentityFile atomically creates <keysDir>/age.txt with mode 0600. The
// target is created with a no-replace hard link from a completed temporary
// file, so a concurrent writer cannot turn this operation into an overwrite.
func WriteIdentityFile(p core.HomerPaths, identity AgeIdentity) error {
	file := IdentityFilePath(p)

	// Check existence before validating the new identity. Even an empty or
	// damaged old file is a reservation that keygen must not silently replace.
	if _, err := os.Lstat(file); err == nil {
		return duplicateIdentityError()
	} else if !errors.Is(err, fs.ErrNotExist) {
		return core.NewCliError("cannot inspect age identity file")
	}

	parsed, err := parseSecretKey(identity.SecretKey)
	if err != nil {
		return err
	}
	serialized := []byte(parsed.String() + "\n")

	dir := keysDirectory(p)
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return core.NewCliError("cannot create age keys directory")
	}
	// MkdirAll's mode is subject to umask and does not change an existing
	// directory. Explicit chmod makes the security invariant hold in both
	// cases.
	if err := os.Chmod(dir, 0o700); err != nil {
		return core.NewCliError("cannot secure age keys directory")
	}

	tmp, err := os.CreateTemp(dir, ".age.txt.tmp-")
	if err != nil {
		return core.NewCliError("cannot create temporary age identity file")
	}
	tmpName := tmp.Name()
	removeTemp := true
	defer func() {
		if removeTemp {
			_ = os.Remove(tmpName)
		}
	}()

	if err := tmp.Chmod(0o600); err != nil {
		_ = tmp.Close()
		return core.NewCliError("cannot secure temporary age identity file")
	}
	if _, err := tmp.Write(serialized); err != nil {
		_ = tmp.Close()
		return core.NewCliError("cannot write age identity file")
	}
	if err := tmp.Sync(); err != nil {
		_ = tmp.Close()
		return core.NewCliError("cannot flush age identity file")
	}
	if err := tmp.Close(); err != nil {
		return core.NewCliError("cannot close age identity file")
	}

	// Link is atomic and refuses to replace an existing target. Both paths are
	// in the same directory, so the operation is on one filesystem.
	if err := os.Link(tmpName, file); err != nil {
		if errors.Is(err, fs.ErrExist) {
			return duplicateIdentityError()
		}
		return core.NewCliError("cannot finalize age identity file")
	}
	if err := os.Remove(tmpName); err != nil {
		return core.NewCliError("cannot finalize age identity file")
	}
	removeTemp = false
	return nil
}

// LoadIdentity reads the local identity. Missing, unreadable, or malformed
// files all return the zero identity and false; callers can decide whether to
// prompt for keygen or report a doctor failure.
func LoadIdentity(p core.HomerPaths) (AgeIdentity, bool) {
	content, err := os.ReadFile(IdentityFilePath(p))
	if err != nil {
		return AgeIdentity{}, false
	}
	identity, err := ParseIdentityFile(string(content))
	if err != nil {
		return AgeIdentity{}, false
	}
	return identity, true
}

// RecipientValid checks the frozen age recipient shape. Cryptographic callers
// additionally parse the recipient with filippo.io/age, which also verifies
// its bech32 checksum.
func RecipientValid(recipient string) bool {
	if len(recipient) != 62 || !strings.HasPrefix(recipient, "age1") {
		return false
	}
	for _, c := range recipient[4:] {
		if !strings.ContainsRune("023456789acdefghjklmnpqrstuvwxyz", c) {
			return false
		}
	}
	return true
}
